File: android/app/src/main/java/com/cochat/app/data/repository/GroupsRepository.kt
package com.cochat.app.data.repository

import com.cochat.app.data.model.GroupInfo
import com.cochat.app.data.remote.AddMembersRequest
import com.cochat.app.data.remote.ApiService
import com.cochat.app.data.remote.CreateGroupRequest
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File

class GroupsRepository(private val api: ApiService) {

    suspend fun create(name: String, memberIds: List<String>): GroupInfo =
        api.createGroup(CreateGroupRequest(name, memberIds)).group

    suspend fun get(id: String): GroupInfo = api.getGroup(id).group

    suspend fun updateName(id: String, name: String): GroupInfo =
        api.updateGroupName(id, name.toRequestBody("text/plain".toMediaType())).group

    suspend fun updateAvatar(id: String, file: File, mimeType: String): GroupInfo {
        val body = file.asRequestBody(mimeType.toMediaType())
        val part = MultipartBody.Part.createFormData("avatar", file.name, body)
        return api.updateGroupAvatar(id, part).group
    }

    suspend fun addMembers(id: String, memberIds: List<String>): GroupInfo =
        api.addMembers(id, AddMembersRequest(memberIds)).group

    suspend fun removeMember(id: String, userId: String) {
        api.removeMember(id, userId)
    }

    suspend fun leave(id: String) {
        api.leaveGroup(id)
    }
}

File: ios/CoChat/Repositories/GroupsRepository.swift
import Foundation

final class GroupsRepository {
    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func create(name: String, memberIds: [String]) async throws -> GroupInfo {
        let res: GroupEnvelope = try await api.post("groups", body: CreateGroupRequest(name: name, memberIds: memberIds))
        return res.group
    }

    func get(id: String) async throws -> GroupInfo {
        let res: GroupEnvelope = try await api.get("groups/\(id)")
        return res.group
    }

    func updateName(id: String, name: String) async throws -> GroupInfo {
        let res: GroupEnvelope = try await api.patchMultipart("groups/\(id)", fields: ["name": name])
        return res.group
    }

    func updateAvatar(id: String, data: Data, fileName: String, mimeType: String) async throws -> GroupInfo {
        let res: GroupEnvelope = try await api.patchMultipart(
            "groups/\(id)", fileField: "avatar", fileData: data, fileName: fileName, mimeType: mimeType
        )
        return res.group
    }

    func addMembers(id: String, memberIds: [String]) async throws -> GroupInfo {
        let res: GroupEnvelope = try await api.post("groups/\(id)/members", body: AddMembersRequest(memberIds: memberIds))
        return res.group
    }

    func removeMember(id: String, userId: String) async throws {
        try await api.deleteVoid("groups/\(id)/members/\(userId)")
    }

    func leave(id: String) async throws {
        try await api.postVoid("groups/\(id)/leave")
    }
}
